File: web/assets/operator-profile.js
/* Fail-closed renderer model for Cordia's non-scored Surveyor operator profile. */
(function (root, factory) {
  var api = factory()
  if (typeof module === 'object' && module.exports) module.exports = api
  root.CordiaOperatorProfile = api
})(typeof globalThis === 'object' ? globalThis : this, function () {
  'use strict'

  var TITLE = 'What Cordia currently understands'
  var ERROR = 'Cordia could not load this profile safely. Try again.'
  var IDENTIFIER = /^[A-Za-z0-9][A-Za-z0-9._-]{0,79}$/
  var CONNECTOR_ID = /^[a-z][a-z0-9_]{0,79}$/
  var CREDENTIAL = /(?:^|[^A-Za-z0-9])(?:sk-|pk-|rk-|gh[pousr]_|github_pat_|xox[baprs]-|AKIA|(?:api[-_.]?key|access[-_.]?token|token|secret|password|authorization|credential)(?:[-_.:=]|\s*=)|bearer\s+|-----BEGIN [A-Z0-9 ]*PRIVATE KEY-----)/i
  var LOCAL_PATH = /(?:^|[^A-Za-z0-9_/.])(?:[A-Za-z]:(?:[\\/]|(?=[^\s\\/]))|\\\\[^\s]+|file:\/\/|path:\/\/|\.{1,2}[\\/][^\s]+|\/(?:tmp|home|Users|var|etc|opt|srv|run|mnt|workspace|Library)(?:\/[^\s]*)?|\/{1,2}(?:[^\s/]+\/)+[^\s/]+)/i
  var URL_CANDIDATE = /\b[A-Za-z][A-Za-z0-9+.-]*:\/\/[^\s<>"']+/g
  var STRENGTH = new Set(['clear', 'emerging', 'early'])
  var ACTION = new Set(['continue_survey', 'refine_profile', 'create_interface'])

  function decoded(value) {
    try { return decodeURIComponent(value) } catch (_) { return null }
  }

  function sensitiveUrlPart(value) {
    var inspected = decoded(value)
    return inspected == null || CREDENTIAL.test(inspected) || LOCAL_PATH.test(inspected)
  }

  function isSensitiveText(value) {
    if (typeof value !== 'string') return false
    if (CREDENTIAL.test(value)) return true
    URL_CANDIDATE.lastIndex = 0
    var spans = []
    var match
    while ((match = URL_CANDIDATE.exec(value))) {
      var url
      try { url = new URL(match[0]) } catch (_) { return true }
      if (url.protocol === 'file:' || url.protocol === 'path:' || !url.hostname ||
          url.username || url.password || sensitiveUrlPart(url.search) || sensitiveUrlPart(url.hash)) return true
      spans.push({ start: match.index, end: match.index + match[0].length })
    }
    var remainder = ''
    var start = 0
    spans.forEach(function (span) {
      remainder += value.slice(start, span.start)
      start = span.end
    })
    remainder += value.slice(start)
    return LOCAL_PATH.test(remainder)
  }

  function safeText(value, limit) {
    if (typeof value !== 'string') return ''
    var text = value.replace(/\s+/g, ' ').trim()
    return text && text.length <= limit && !isSensitiveText(text) ? text : ''
  }

  function safeIdentifier(value) {
    return typeof value === 'string' && IDENTIFIER.test(value) && !isSensitiveText(value)
  }

  function identifiers(items) {
    return (Array.isArray(items) ? items : []).slice(0, 3).flatMap(function (item) {
      if (!item || typeof item !== 'object') return []
      var name = safeText(item.name, 80)
      var meaning = safeText(item.meaning, 240)
      var useAi = safeText(item.use_ai_this_way, 240)
      if (!name || !meaning || !useAi || !STRENGTH.has(item.evidence_strength)) return []
      return [{ name: name, meaning: meaning, useAiThisWay: useAi, evidenceStrength: item.evidence_strength }]
    })
  }

  function labelled(items) {
    return (Array.isArray(items) ? items : []).slice(0, 6).flatMap(function (item) {
      if (!item || typeof item !== 'object') return []
      var label = safeText(item.label, 80)
      var value = safeText(item.value, 240)
      return label && value ? [{ label: label, value: value }] : []
    })
  }

  function evidence(items) {
    return (Array.isArray(items) ? items : []).slice(0, 6).flatMap(function (item) {
      if (!item || typeof item !== 'object') return []
      var summary = safeText(item.summary, 280)
      return summary && STRENGTH.has(item.evidence_strength)
        ? [{ summary: summary, evidenceStrength: item.evidence_strength }]
        : []
    })
  }

  function connectors(items) {
    return (Array.isArray(items) ? items : []).slice(0, 12).flatMap(function (item) {
      if (!item || typeof item !== 'object') return []
      var name = safeText(item.name, 80)
      if (!CONNECTOR_ID.test(item.id || '') || !name ||
          !['Confirmed by user', 'Suggested - not connected'].includes(item.status) ||
          !['live', 'planned'].includes(item.implementation_status)) return []
      return [{ id: item.id, name: name, status: item.status, implementationStatus: item.implementation_status }]
    })
  }

  function learning(items) {
    return (Array.isArray(items) ? items : []).slice(0, 5)
      .map(function (item) { return safeText(item, 160) }).filter(Boolean)
  }

  function nextAction(item) {
    if (!item || typeof item !== 'object' || !ACTION.has(item.type)) return null
    var label = safeText(item.label, 80)
    var reason = safeText(item.reason, 240)
    return label && reason ? { type: item.type, label: label, reason: reason } : null
  }

  function workspace(item, navigation) {
    if (!item || typeof item !== 'object' || !safeIdentifier(item.id) || !navigation ||
        typeof navigation.buildWorkspaceNavigation !== 'function') return null
    var target = navigation.buildWorkspaceNavigation(item.id)
    var name = safeText(item.name, 80)
    return target && name ? { id: item.id, name: name, href: target.href } : null
  }

  function errorModel() {
    return { state: 'error', message: ERROR }
  }

  function buildOperatorProfileModel(payload, navigation) {
    if (!payload || payload.ok !== true || !payload.operator_profile ||
        typeof payload.operator_profile !== 'object') return errorModel()
    var source = payload.operator_profile
    if (source.title !== TITLE || isSensitiveText(source.title)) return errorModel()
    var action = nextAction(source.next_action)
    if (!action) return errorModel()
    var latest = workspace(source.latest_workspace, navigation)
    var primary
    var secondary = null
    if (latest) {
      primary = { kind: 'link', label: 'Open ' + latest.name, href: latest.href }
      if (action.type !== 'create_interface') secondary = { kind: 'surveyor', label: 'Refine with Surveyor' }
    } else if (action.type === 'create_interface') {
      primary = { kind: 'link', label: 'Build this workspace', href: 'builder.html?from=surveyor' }
    } else {
      primary = { kind: 'surveyor', label: action.label }
    }
    return {
      state: 'ready',
      title: TITLE,
      identifiers: identifiers(source.identifiers),
      understanding: labelled(source.understanding),
      evidence: evidence(source.evidence),
      connectors: connectors(source.connectors),
      stillLearning: learning(source.still_learning),
      nextAction: action,
      primaryAction: primary,
      secondaryAction: secondary,
    }
  }

  return { buildOperatorProfileModel: buildOperatorProfileModel, isSensitiveText: isSensitiveText }
})

File: backend/surveyor/operator_profile.py
"""Least-privilege, non-scored projection for the Surveyor operator profile."""
from __future__ import annotations

import re
from urllib.parse import unquote, urlsplit

from . import artifacts, freeform, identifiers, scenarios, types


_TEXT_LIMIT = 320
_IDENTIFIER = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,79}$")
_CONNECTOR_ID = re.compile(r"^[a-z][a-z0-9_]{0,79}$")
_CREDENTIAL = re.compile(
    r"(?:^|[^A-Za-z0-9])(?:sk-|pk-|rk-|gh[pousr]_|github_pat_|xox[baprs]-|"
    r"AKIA|(?:api[-_.]?key|access[-_.]?token|token|secret|password|authorization|"
    r"credential)(?:[-_.:=]|\s*=)|bearer\s+|"
    r"-----BEGIN [A-Z0-9 ]*PRIVATE KEY-----)", re.IGNORECASE)
_LOCAL_PATH = re.compile(
    r"(?:^|[^A-Za-z0-9_/.])(?:[A-Za-z]:(?:[\\/]|(?=[^\s\\/]))|"
    r"\\\\[^\s]+|(?:file|path)://|\.{1,2}[\\/][^\s]+|"
    r"/(?:tmp|home|Users|var|etc|opt|srv|run|mnt|workspace|Library)(?:/[^\s]*)?|"
    r"/{1,2}(?:[^\s/]+/)+[^\s/]+|"
    r"(?:private|secret|credentials?|keys?)(?:[\\/][^\s]+)+|"
    r"[A-Za-z0-9_.-]+(?:[\\/][A-Za-z0-9_.-]+)*[\\/][A-Za-z0-9_-]+\.[A-Za-z0-9]{1,10}(?:[\\/][^\s]*)?|"
    r"/[A-Za-z0-9_-]+\.[A-Za-z0-9]{1,10})", re.IGNORECASE)
_REMOTE_URL = re.compile(r"\b[A-Za-z][A-Za-z0-9+.-]*://[^\s<>\"']+")
_BAD_PERCENT = re.compile(r"%(?![0-9A-Fa-f]{2})")
_STRENGTH = {
    "clear": "clear", "high": "clear",
    "emerging": "emerging", "medium": "emerging",
    "early": "early", "low": "early",
}
_ACTION_TYPES = {"continue_survey", "refine_profile", "create_interface"}


def build(profile, connector_states=None, interfaces=None):
    """Project existing canonical state without adding another state owner."""
    profile = profile if isinstance(profile, dict) else {}
    connector_states = artifacts.normalize_connector_states(
        connector_states if isinstance(connector_states, dict) else {})
    assessment_profile = _validated_assessment_profile(profile)
    assessment = artifacts.assessment_view(assessment_profile, connector_states)
    return {
        "title": "What Cordia currently understands",
        "identifiers": _identifiers(profile.get("identifiers")),
        "understanding": _understanding(assessment.get("understanding")),
        "evidence": _evidence(assessment.get("evidence")),
        "connectors": _connectors(assessment.get("connectors")),
        "still_learning": _still_learning(assessment_profile, connector_states),
        "next_action": _next_action(
            _validated_action_profile(profile), complete=is_complete(profile)),
        "latest_workspace": _latest_workspace(interfaces),
    }


def public_identifiers(items):
    """Allow-list positive display fields for every browser-facing profile."""
    return _identifiers(items)


def next_action(profile):
    """Return the existing forward action after validating legacy profile shapes."""
    profile = profile if isinstance(profile, dict) else {}
    return _next_action(
        _validated_action_profile(profile), complete=is_complete(profile))


def is_complete(profile):
    """Use the canonical three-stage completion rule on bounded known keys only."""
    profile = profile if isinstance(profile, dict) else {}
    raw_scenarios = profile.get("scenarios")
    raw_freeform = profile.get("freeform")
    completion_profile = {
        "questions_answered": profile.get("questions_answered"),
        "signals": types.validate_signals(profile.get("signals")),
        "scenarios": {
            key: value for key, value in (raw_scenarios.items()
                                           if isinstance(raw_scenarios, dict) else [])
            if key in scenarios.IDS
        },
        "freeform": {
            key: value for key, value in (raw_freeform.items()
                                           if isinstance(raw_freeform, dict) else [])
            if key in freeform.KEYS and freeform.clean(value)
        },
    }
    return types.onboarding_complete(completion_profile)


def _validated_assessment_profile(profile):
    evidence = []
    for item in profile.get("evidence") if isinstance(profile.get("evidence"), list) else []:
        if not isinstance(item, dict):
            continue
        summary = item.get("summary")
        confidence = item.get("confidence")
        if isinstance(summary, str) and isinstance(confidence, str):
            evidence.append({"summary": summary, "confidence": confidence})
    return {
        "signals": types.validate_signals(profile.get("signals")),
        "evidence": evidence[:40],
    }


def _validated_action_profile(profile):
    return {
        "signals": types.validate_signals(profile.get("signals")),
        "scores": types.validate_scores(profile.get("scores")),
        "evidence": types.validate_evidence(profile.get("evidence")),
    }


def _identifiers(items):
    output = []
    for item in items if isinstance(items, list) else []:
        if not isinstance(item, dict):
            continue
        name = _safe_text(item.get("name"), 80)
        meaning = _safe_text(item.get("meaning"), 240)
        use_ai = _safe_text(item.get("use_ai_this_way"), 240)
        strength = _strength(item.get("confidence"))
        if name and meaning and use_ai and strength:
            output.append({
                "name": name,
                "meaning": meaning,
                "use_ai_this_way": use_ai,
                "evidence_strength": strength,
            })
        if len(output) == 3:
            break
    return output


def _understanding(items):
    output = []
    for item in items if isinstance(items, list) else []:
        if not isinstance(item, dict):
            continue
        label = _safe_text(item.get("label"), 80)
        value = _safe_text(item.get("value"), 240)
        if label and value:
            output.append({"label": label, "value": value})
        if len(output) == 6:
            break
    return output


def _evidence(items):
    output = []
    for item in items if isinstance(items, list) else []:
        if not isinstance(item, dict):
            continue
        summary = _safe_text(item.get("summary"), 280)
        strength = _strength(item.get("confidence"))
        if summary and strength:
            output.append({"summary": summary, "evidence_strength": strength})
        if len(output) == 6:
            break
    return output


def _connectors(items):
    output = []
    for item in items if isinstance(items, list) else []:
        if not isinstance(item, dict):
            continue
        connector_id = item.get("id")
        name = _safe_text(item.get("name"), 80)
        status = item.get("status")
        implementation = item.get("implementation_status")
        if (isinstance(connector_id, str) and _CONNECTOR_ID.fullmatch(connector_id) and name and
                status in {"Confirmed by user", "Suggested - not connected"} and
                implementation in {"live", "planned"}):
            output.append({
                "id": connector_id,
                "name": name,
                "status": status,
                "implementation_status": implementation,
            })
        if len(output) == 12:
            break
    return output


def _strength(value):
    return _STRENGTH.get(value) if isinstance(value, str) else None


def _still_learning(profile, connector_states):
    signals = profile.get("signals") if isinstance(profile.get("signals"), dict) else {}
    items = []
    if not _safe_text(signals.get("domain"), 240):
        items.append("Your work domain and operating context")
    if not _safe_text(signals.get("primary_goal"), 240):
        items.append("The first outcome you want Cordia to improve")
    if signals.get("delegation_style") not in {
            "agent_autonomous", "human_checkpoint_before_final", "human_reviews_every_step"}:
        items.append("Where you want a human checkpoint")
    if not _evidence(profile.get("evidence")):
        items.append("Examples in your own words")
    if not artifacts.normalize_connector_states(connector_states):
        items.append("Which systems should be connected")
    return items[:5]


def _next_action(profile, complete=False):
    if complete:
        return {
            "type": "create_interface",
            "label": "Build my workspace",
            "reason": "Your Surveyor intake is complete and ready to shape a workspace.",
        }
    action_profile = dict(profile)
    action_profile["scores"] = (profile.get("scores")
                                if isinstance(profile.get("scores"), dict) else {})
    action_profile["signals"] = (profile.get("signals")
                                 if isinstance(profile.get("signals"), dict) else {})
    action = identifiers.next_best_action(action_profile)
    action_type = action.get("type")
    if action_type == "create_interface":
        action = {
            "type": "refine_profile",
            "label": "Continue with Surveyor",
            "reason": "Finish the bounded intake before Cordia builds your workspace.",
        }
        action_type = action["type"]
    label = _safe_text(action.get("label"), 80)
    reason = _safe_text(action.get("reason"), 240)
    if action_type not in _ACTION_TYPES or not label or not reason:
        return {
            "type": "continue_survey",
            "label": "Talk to Surveyor",
            "reason": "A few answers help Cordia shape the right workspace.",
        }
    return {"type": action_type, "label": label, "reason": reason}


def _latest_workspace(interfaces):
    if not isinstance(interfaces, list) or not interfaces:
        return None
    latest = interfaces[0]
    if not isinstance(latest, dict):
        return None
    workspace_id = latest.get("id")
    if not _safe_identifier(workspace_id):
        return None
    return {
        "id": workspace_id,
        "name": _safe_text(latest.get("name"), 80) or "Workspace",
    }


def _safe_identifier(value):
    return (isinstance(value, str) and bool(_IDENTIFIER.fullmatch(value)) and
            not _is_sensitive(value))


def _safe_text(value, limit=_TEXT_LIMIT):
    if not isinstance(value, str):
        return ""
    value = " ".join(value.split()).strip()
    if not value or len(value) > limit or _is_sensitive(value):
        return ""
    return value


def _is_sensitive(value):
    if _CREDENTIAL.search(value):
        return True
    spans = []
    for match in _REMOTE_URL.finditer(value):
        try:
            remote = match.group(0)
            if _BAD_PERCENT.search(remote):
                return True
            parsed = urlsplit(remote)
            if (parsed.scheme.lower() in {"file", "path"} or not parsed.hostname or
                    parsed.username or parsed.password):
                return True
            details = unquote((parsed.query or "") + " " + (parsed.fragment or ""))
            if _CREDENTIAL.search(details) or _LOCAL_PATH.search(details):
                return True
            spans.append(match.span())
        except (TypeError, ValueError, UnicodeError):
            return True
    remainder = []
    start = 0
    for left, right in spans:
        remainder.append(value[start:left])
        start = right
    remainder.append(value[start:])
    return bool(_LOCAL_PATH.search("".join(remainder)))
